Fix Application Passwords login issue

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11649 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
We identify an issue with login where the fact that `SelectedSite`'s operations are not thread-safe can break the login flow, for more information on the exact flow where this happens, please check this [comment](https://github.com/woocommerce/woocommerce-android/issues/11649#issuecomment-2145811408).
This PR fixes the issue by making `SelectedSite` thread safe.

### Steps to reproduce
1. Check out `trunk`.
2. Apply the patch from below (the patch just attempts to increase the chance of the race condition)
3. Launch the app then sign in using Application Passwords.
4. Notice that the login doesn't work, the prologue screen is shown after the successful login.

**You might still need to repeat (3) few times to hit the issue.**

### Testing information
##### Confirm the issue is fixed
1. Check out the branch of this PR.
2. Apply the patch from below (the patch just attempts to increase the chance of the race condition)
3. Launch the app then sign in using Application Passwords.
4. Confirm the login works without issues.
5. Repeat few times to confirm the issue is fixed.

##### Non-regression
1. Smoke test the app.
2. Smoke test the site picker.

<details>
<summary>
<h3>Patch</h3>
</summary>

```patch
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt	(revision faa398d1f8c6015b4102f02fed0b800b6a53fb44)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt	(date 1717438482298)
@@ -30,7 +30,9 @@
 import com.woocommerce.android.viewmodel.getNullableStateFlow
 import com.woocommerce.android.viewmodel.getStateFlow
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.combine
 import kotlinx.coroutines.flow.flatMapLatest
@@ -49,6 +51,7 @@
 import org.wordpress.android.util.UrlUtils
 import java.net.URI
 import javax.inject.Inject
+import kotlin.random.Random
 
 @HiltViewModel
 class LoginSiteCredentialsViewModel @Inject constructor(
@@ -272,6 +275,7 @@
                     when (authenticationError?.errorType) {
                         GENERIC_ERROR,
                         INVALID_CREDENTIALS -> errorDialogMessage.value = authenticationError.errorMessage
+
                         else -> {
                             fetchSiteForTutorial(
                                 username = state.username,
@@ -390,7 +394,13 @@
                     loginAnalyticsListener.trackAnalyticsSignIn(false)
                 }
                 appPrefs.removeLoginSiteAddress()
+                repeat(10) {
+                    launch(Dispatchers.Default) {
+                        selectedSite.getOrNull()
+                    }
+                }
                 selectedSite.set(site)
+
                 triggerEvent(LoggedIn(selectedSite.getSelectedSiteId()))
             },
             onFailure = { exception ->
```

</details>

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Make SelectedSite thread safe

We simply synchronize access to its functions, this will allow avoiding cases where the selected site ID is cleared due to race conditions.
Check: https://github.com/woocommerce/woocommerce-android/issues/11649#issuecomment-2145811408

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  1 Message: This PR contains changes to <c...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/217">18.9 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR contains changes to <code>RELEASE-NOTES.txt</code>.<br />
Note that these changes won't affect the final version of the release notes as this version is in code freeze.<br />
Please, get in touch with a release manager if you want to update the final release notes.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/tools/SelectedSite.kt:
```diff
@@ -63,26 +63,29 @@ class SelectedSite @Inject constructor(
     fun get(): SiteModel {
         state.value?.let { return it }
```
I kept this outside of the synchronized block because it's not involved on any of the changes that could lead to the issue, and also to avoid any negative impact on the app's performance, as this code is what will be executed in almost all of the app's lifetime after selecting a site.

JorgeMucientes commented:
Great job, @hichamboushaba. This was a really tricky race condition to catch! The synchronized blocks are exactly where they need to be, and IMHO, the changes are safe enough to merge this as a beta fix. 

Ready for merging once we decide the target branch.

hichamboushaba committed:
Move function to group synchronized functions together

hichamboushaba commented:
Thanks @JorgeMucientes for the review, sounds good, let's target 18.9 with this, I rebased the branch, and I added one last commit 74ac90b to move the function `reset` near the `set` function just for a better organization of the code, and to make it easy to see the synchronized functions together.

hichamboushaba committed:
Add a release notes entry
